### PR TITLE
Remove old invoker code and refactor accordingly.

### PR DESCRIPTION
**Pull request by markusthoemmes**
**Note:** This is not to be merged before the 16th of August!

- Refactoring the invoker for better encapsulation and seperation of concerns in its packages.
- Removing Dispatcher.

TODO:

- [x] Remove unused properties (useRunC)

### CODE REVIEW

File: core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala [hunk 1]
```diff
@@ -147,70 +156,74 @@ class InvokerReactive(
...
-        }
+
+                WhiskAction.get(entityStore, actionid.id, actionid.rev, fromCache = actionid.rev != DocRevision.empty).flatMap { action =>
+                    action.toExecutableWhiskAction match {
+                        case Some(executable) =>
+                            pool ! Run(executable, msg)
+                            Future.successful(())
+                        case None =>
+                            logging.error(this, s"non-executable action reached the invoker ${action.fullyQualifiedName(false)}")
+                            Future.failed(new IllegalStateException())
```
Comment (satybald) [inline core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: what do you think about adding proper message to the IllegalStateException?

**Comment by markusthoemmes** [comment]
PG3 1010 🔵

File: core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala (hunk 1 above)
Comment (markusthoemmes) [inline core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: We could do that for good measure. Note though that the message wouldn't be displayed anywhere (at least not today)

File: core/invoker/src/main/scala/whisk/core/dispatcher/Dispatcher.scala [hunk 2]
```diff
@@ -1,159 +0,0 @@
...
-     * There is typically only one handler.
-     */
-    def process(bytes: Array[Byte]): Future[Unit] = Future {
-        val raw = new String(bytes, StandardCharsets.UTF_8)
-        ActivationMessage.parse(raw) match {
-            case Success(m) =>
-                handlers foreach {
-                    case (name, handler) => handleMessage(handler, m)
-                }
-            case Failure(t) => logging.info(this, errorMsg(raw, t))
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/dispatcher/Dispatcher.scala]: only real change i see is that we lose this log message should a message fail to parse.

File: core/invoker/src/main/scala/whisk/core/dispatcher/Dispatcher.scala [hunk 3]
```diff
@@ -1,159 +0,0 @@
...
-    private def handleMessage(handler: MessageHandler, msg: ActivationMessage): Unit = {
-        implicit val tid = msg.transid
-
-        Future {
-            val count = counter.next()
-            logging.debug(this, s"activeCount = $count while handling ${handler.name}")
-            handler.onMessage(msg) // returns a future which is flat-mapped via identity to hang onComplete
-        } flatMap (identity) onComplete {
-            case Success(a) => logging.debug(this, s"activeCount = ${counter.prev()} after handling ${handler.name}")
-            case Failure(t) => logging.error(this, s"activeCount = ${counter.prev()} ${errorMsg(handler, t)}")
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/dispatcher/Dispatcher.scala]: and related to the IllegalStateException below, we no longer log the failed future from the invoker.

File: core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala [hunk 4]
```diff
@@ -222,7 +222,15 @@ class InvokerReactive(config: WhiskConfig, instance: InstanceId, producer: Messa
                         activationFeed ! MessageFeed.Processed
                         ack(msg.transid, activation, msg.rootControllerIndex)
                         store(msg.transid, activation)
+                        Future.successful(())
                 }
+            }.recoverWith {
+                case t =>
+                    // Iff everything above failed, we have a terminal error at hand. Either the message failed
+                    // to deserialize, or something threw an error where it is not expected to throw.
+                    activationFeed ! MessageFeed.Processed
```
Comment (rabbah) [inline core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala]: good catch.

**Pull request merged**